genrl/core/actor_critic.py

```diff
@@ -41,6 +41,73 @@ def __init__(
         self.critic = MlpValue(state_dim, action_dim, val_type, value_layers, **kwargs)
 
 
+class MlpSharedActorCritic(BaseActorCritic):
+    """MLP Shared Actor Critic
+
+    Attributes:
+        state_dim (int): State dimensions of the environment
+        action_dim (int): Action space dimensions of the environment
+        hidden (:obj:`list` or :obj:`tuple`): Hidden layers in the MLP
+        val_type (str): Value type of the critic network
+        discrete (bool): True if the action space is discrete, else False
+        sac (bool): True if a SAC-like network is needed, else False
+        activation (str): Activation function to be used. Can be either "tanh" or "relu"
+    """
+
+    def __init__(
+        self,
+        state_dim: spaces.Space,
+        action_dim: spaces.Space,
+        shared_layers: Tuple = (32, 32),
+        policy_layers: Tuple = (32, 32),
+        value_layers: Tuple = (32, 32),
+        val_type: str = "V",
+        discrete: bool = True,
+        **kwargs,
+    ):
+        super(MlpSharedActorCritic, self).__init__()
+        self.shared = mlp([state_dim] + list(shared_layers))
+        self.actor = MlpPolicy(
+            shared_layers[-1], action_dim, policy_layers, discrete, **kwargs
+        )
+        self.critic = MlpValue(
+            shared_layers[-1], action_dim, val_type, value_layers, **kwargs
+        )
+        self.state_dim = state_dim
+        self.action_dim = action_dim
+
+    def get_features(self, state: torch.Tensor):
+        features = self.shared(state)
+        return features
+
+    def get_action(self, state: torch.Tensor, deterministic: bool = False):
+        state = torch.as_tensor(state).float()
+        features = self.get_features(state)
+        action_probs = self.actor(features)
+        action_probs = nn.Softmax(dim=-1)(action_probs)
+
+        if deterministic:
+            action = torch.argmax(action_probs, dim=-1).unsqueeze(-1).float()
+            distribution = None
+        else:
+            distribution = Categorical(probs=action_probs)
+            action = distribution.sample()
+
+        return action, distribution
+
+    def get_value(self, state: torch.Tensor):
+        state = torch.as_tensor(state).float()
+        if self.critic.val_type == "Qsa":
+            features = self.shared(state[:, :, :-1])
+            features = torch.cat([features, state[:, :, -1].unsqueeze(-1)], dim=-1)
+            print(f"features {features.shape}")
+            value = self.critic(features).float().squeeze(-1)
+        else:
+            features = self.shared(state)
+            value = self.critic(features)
```

Comment:
Why the need for the special case? What error are you getting here?

Reply:
> Why the need for the special case? What error are you getting here?

There's a size mismatch that arises in "Qsa" since the Mlpvalue would still take in [state_dim + action_dim] as input
So it made sense to process the state first and input the action just before the value head diverged from the shared layers

Reply:
Does this problem happen with the normal Mlp too?

Reply:
it doesnt arise there - maybe this code example might help visualise it better
```
# normal mlp
self.critic = MlpValue(state_dim, action_dim, "Qsa", value_layers, **kwargs)
# input dimensions = [state_dim + action_dim]

def get_value(self, state):
    # Here state is passed as torch.cat([state, action])
    value = self.critic(state)
    return value

# Shared layers
self.critic = MlpValue(shared_layers[-1], action_dim, "Qsa", value_layers, **kwargs)
# input_dimensions = [shared_layers[-1] + action_dim]

def get_value(self, state):
    # Here state is passed as torch.cat([state, action])
    x = state[:, :, :-1] # extract just the state
    a = state[:, :, -1].unsqueeze(-1)
    features = self.features(state) # output_dimenstions = [shared_layers[-1]]
    state = torch.cat([state, a]) # output_dimensions = [shared_layers[-1] + action_dim]

    value = self.critic(features)
    return value

#You could also do this with the normal mlp
self.critic = MlpValue(state_dim, action_dim, "Qsa", value_layers, **kwargs)
# input dimensions = [state_dim + action_dim]

def get_value(self, state):
    # Here state is passed as torch.cat([state, action])
    x = state[:, :, :-1] # extract just the state
    a = state[:, :, -1].unsqueeze(-1) # extract the actions
    new_state = torch.cat([x, a])
    value = self.critic(new_state) # but new_state is the same as state, hence it works fine
    return value
```

Reply:
Though more accurately, it should be `state[:, :, :-action_dim]` in place of `state[:, :, :-1]`